File: src/causingToolCall.ts
// src/causingToolCall.ts — given a burn PEAK (a time window, plus a session or workspace), extract
// EVERY spawn tool-call that could have caused it, VERBATIM, from the Claude session JSONL — numbered,
// time-ordered, and measured (composition counts). NOT a single "nearest" call: a fork-storm is a
// SUSTAINED burst driven by potentially many spawns (Workflows that fan out internally, repeated
// Agent/Task calls), so reporting one is misleading — the full list is the honest view.
//
// WHY: a burn peak with no causing calls is useless — the operator still has to hand-dig the
// transcript. Every peak reporter (investigate_burn, --risk / burn-guard, watch, get_burn_status)
// attaches this so the peak lists WHAT spawned the storm, when, and how many of each.
//
// HOW: transcripts reach ~2 GB (base64 images), so we do NOT parse them in JS. DuckDB's read_json —
// already a runtime dep, it backs the forensics store — streams the NDJSON in C++, projects only
// {timestamp,type,message}, and extracts every spawn tool_use in the window with a JSON path. Reads
// the JSONL (always present), so it works even when raw-body capture is OFF. maximum_object_size is
// raised past the largest image-bloated line so a fat line is SKIPPED (ignore_errors), never aborting
// the read. Query verified against a live transcript before shipping.
//
// Fast-path discipline: callers invoke this ONLY when a peak actually fires (never on the quiet
// path), so --risk's ~50 ms budget is untouched; the streaming read is paid only to explain a peak.

import * as fs from 'fs'
import * as path from 'path'
import { claudeProjectsDirs } from './logReader'
import { resolveProjectSlugs } from './projectSlug'
import { transcriptReadSpec, tornLineSql } from './ndjsonDuck'

/** Tools whose call fans out / spawns work — the candidates behind a FORK_STORM / fan-out burst. */
export const SPAWN_TOOLS = ['Task', 'Agent', 'Workflow', 'SendMessage'] as const

/** One spawn tool-call found in the window. */
export interface SpawnCall {
  /** 1-based index in TIME order (oldest first) — the "numbered" view. */
  n: number
  /** When the call was issued (ISO). */
  iso: string
  /** The tool, e.g. 'Workflow' | 'Agent'. */
  tool: string
  /** input.subagent_type when present (Agent/Task), else null. */
  subagentType: string | null
  /** input.model when the call pinned one, else null (inherited). */
  model: string | null
  /** The tool_use input, VERBATIM (JSON) — the exact call, not a digest. */
  input: string
  /** The session whose transcript issued the call. */
  sessionId: string
}

export interface CausingCallsOptions {
  /** The peak time. Spawns are collected across [atMs - windowMs, atMs + forwardSlackMs]. */
  atMs: number
  /** Live peaks know it (SubagentStart payload). Resolves <sessionId>.jsonl directly. */
  sessionId?: string
  /** investigate_burn knows the workspace — resolves that workspace's transcripts. */
  workspace?: string
  /** Explicit transcript file (tests / a known path). Takes precedence. */
  jsonlPath?: string
  /** How far back before atMs to collect spawns (a sustained burst spans time). Default 15 min. */
  windowMs?: number
  /** Allowance AFTER atMs for clock skew. Default 90 s. */
  forwardSlackMs?: number
  /** Spawn-tool filter. Default SPAWN_TOOLS. */
  tools?: readonly string[]
  /** Override claudeProjectsDirs() — tests only. */
  projectsDirs?: string[]
}

export type CausingCallReason = 'no-locator' | 'no-transcript' | 'none-in-window' | 'duckdb-unavailable'

export interface CausingCallsResult {
  /** Every spawn call in the window, oldest-first (numbered). Empty when none / unresolved. */
  calls: SpawnCall[]
  windowFromIso: string
  windowToIso: string
  /** Set when `calls` is empty — the honest reason, never a fabricated call. */
  reason?: CausingCallReason
  /** Set only when the scanned transcript(s) contained unparseable (torn) NDJSON lines — the
   *  human-readable disclosure that ignore_errors silently degraded rather than dropped them. */
  note?: string
}

const DEFAULT_WINDOW_MS = 15 * 60_000
const DEFAULT_FORWARD_SLACK_MS = 90_000
// mtime is a coarse "this session was active near the peak" proxy — a session file spans time, so
// widen the mtime gate by an hour on each side before the precise ts filter runs inside DuckDB.
const MTIME_SLACK_MS = 3600_000
const MAX_FILES = 8


/** Candidate transcript file(s) for this peak, most-recently-modified first (capped). */
function resolveTranscripts(opts: CausingCallsOptions): string[] {
  if (opts.jsonlPath) return fs.existsSync(opts.jsonlPath) ? [opts.jsonlPath] : []
  const bases = opts.projectsDirs ?? claudeProjectsDirs()

  if (opts.sessionId) {
    const out: string[] = []
    for (const base of bases) {
      let subs: string[]
      try { subs = fs.readdirSync(base) } catch { continue }
      for (const sub of subs) {
        const p = path.join(base, sub, `${opts.sessionId}.jsonl`)
        if (fs.existsSync(p)) out.push(p)
      }
    }
    return out
  }

  if (opts.workspace) {
    // Resolved against disk rather than derived: Claude Code truncates-and-hashes a slug past 200
    // chars, so a deep workspace path produces a directory name no formula here can predict — and
    // the naive derivation names a directory that cannot exist, finding no transcripts at all.
    const slugs = resolveProjectSlugs(opts.workspace, bases)
    const lo = opts.atMs - (opts.windowMs ?? DEFAULT_WINDOW_MS) - MTIME_SLACK_MS
    const hi = opts.atMs + (opts.forwardSlackMs ?? DEFAULT_FORWARD_SLACK_MS) + MTIME_SLACK_MS
    const cand: { p: string; mtime: number }[] = []
    for (const base of bases) for (const slug of slugs) {
      const dir = path.join(base, slug)
      let names: string[]
      try { names = fs.readdirSync(dir) } catch { continue }
      for (const n of names) {
        if (!n.endsWith('.jsonl')) continue
        const p = path.join(dir, n)
        try {
          const st = fs.statSync(p)
          if (st.mtimeMs < lo || st.mtimeMs > hi) continue
          cand.push({ p, mtime: st.mtimeMs })
        } catch { /* raced/unreadable — skip */ }
      }
    }
    return cand.sort((a, b) => b.mtime - a.mtime).slice(0, MAX_FILES).map(c => c.p)
  }

  return []
}

const iso = (ms: number): string => new Date(ms).toISOString()
/** Single-quote a value for inlining in SQL. Paths/tool names are machine-local, not user input,
 *  but escaping the quote keeps a path with an apostrophe from breaking (or injecting) the query. */
const sqlStr = (s: string): string => `'${s.replace(/'/g, "''")}'`

/**
 * EVERY spawn tool-call in the peak window, numbered + time-ordered + verbatim — or an empty list
 * with a reason (never fabricated). One DuckDB streaming pass over the candidate transcript(s).
 */
export async function causingToolCalls(opts: CausingCallsOptions): Promise<CausingCallsResult> {
  const windowMs = opts.windowMs ?? DEFAULT_WINDOW_MS
  const fwd = opts.forwardSlackMs ?? DEFAULT_FORWARD_SLACK_MS
  const windowFromIso = iso(opts.atMs - windowMs)
  const windowToIso = iso(opts.atMs + fwd)

  const files = resolveTranscripts(opts)
  if (files.length === 0) {
    const located = Boolean(opts.jsonlPath || opts.sessionId || opts.workspace)
    return { calls: [], windowFromIso, windowToIso, reason: located ? 'no-transcript' : 'no-locator' }
  }

  let duck: typeof import('@duckdb/node-api')
  try { duck = await import('@duckdb/node-api') } catch { return { calls: [], windowFromIso, windowToIso, reason: 'duckdb-unavailable' } }

  const tools = opts.tools ?? SPAWN_TOOLS
  const toolList = tools.map(sqlStr).join(', ')

  // filename=true tags each row with its source file so a multi-session workspace scan knows WHICH
  // session issued each call. json_extract(... '$.content[*]') returns a LIST that UNNEST flattens to
  // one row per content block; ORDER BY ts ASC gives the numbered, oldest-first burst timeline.
  const readJson = transcriptReadSpec(files)
  const sql = `
    WITH lines AS (
      SELECT filename, timestamp, type, message
      FROM ${readJson}
      WHERE type='assistant' AND timestamp >= ${sqlStr(windowFromIso)} AND timestamp <= ${sqlStr(windowToIso)}
    ),
    blocks AS (
      SELECT filename, timestamp, UNNEST(json_extract(message, '$.content[*]')) AS block FROM lines
    )
    SELECT filename, timestamp AS ts,
           json_extract_string(block, '$.name') AS tool,
           json_extract_string(block, '$.input.subagent_type') AS subagent_type,
           json_extract_string(block, '$.input.model') AS model,
           CAST(json_extract(block, '$.input') AS VARCHAR) AS input
    FROM blocks
    WHERE json_extract_string(block, '$.type') = 'tool_use'
      AND json_extract_string(block, '$.name') IN (${toolList})
    ORDER BY timestamp ASC;`

  const inst = await duck.DuckDBInstance.create(':memory:')
  const con = await inst.connect()
  try {
    const rows = (await con.runAndReadAll(sql)).getRowObjects()

    // Torn-line disclosure: only after the main query succeeded (one extra round-trip). A malformed
    // line lands as an all-NULL row under ignore_errors, not a dropped one — count(*) alone would
    // pass silently, so compare it against a column EVERY real record carries.
    //
    // That column is `type`, NOT `timestamp`. Measured over 482,993 real transcript records:
    // `type` is missing from 0 of them, `timestamp` from 81,814 — 16.9%, because `attachment`,
    // `queue-operation` and `last-prompt` records legitimately carry no timestamp. This probe reads
    // the UNFILTERED scan (the `type='assistant'` filter belongs to the query above, not here), so
    // keying on `timestamp` would have reported roughly a sixth of a healthy machine's records as
    // "unparseable and excluded" — a disclosure that lies, which is worse than none at all.
    const [{ total: tlTotal, withCol: tlWithCol }] =
      (await con.runAndReadAll(tornLineSql(readJson, 'type'))).getRowObjects()
        .map(r => ({ total: Number(r.total ?? 0), withCol: Number(r.withCol ?? 0) }))
    const tornLines = tlTotal - tlWithCol
    const note = tornLines > 0 ? `${tornLines} line(s) unparseable and excluded.` : undefined

    if (rows.length === 0) return { calls: [], windowFromIso, windowToIso, reason: 'none-in-window', note }
    const calls: SpawnCall[] = rows.map((r, i) => ({
      n: i + 1,
      iso: String(r.ts),
      tool: String(r.tool),
      subagentType: r.subagent_type == null ? null : String(r.subagent_type),
      model: r.model == null ? null : String(r.model),
      input: String(r.input),
      sessionId: opts.sessionId ?? path.basename(String(r.filename), '.jsonl'),
    }))
    return { calls, windowFromIso, windowToIso, note }
  } finally {
    con.closeSync()
    inst.closeSync()
  }
}

/** Compact "Agent/general-purpose/opus×2, Workflow×2, …" tally of a spawn list, most-frequent first. */
export function composition(calls: SpawnCall[]): string {
  const by = new Map<string, number>()
  for (const c of calls) {
    const key = [c.tool, c.subagentType, c.model].filter(Boolean).join('/')
    by.set(key, (by.get(key) ?? 0) + 1)
  }
  return [...by.entries()].sort((a, b) => b[1] - a[1]).map(([k, n]) => `${k}×${n}`).join(', ')
}

/** Numbered, time-ordered, measured block for text reporters (--risk, watch, investigate_burn).
 *  Full literal input per call (the user's explicit choice); an empty result renders its honest
 *  reason, never a guess. */
export function renderCausingCalls(r: CausingCallsResult): string {
  const noteSuffix = r.note ? ` ${r.note}` : ''
  if (r.calls.length === 0) return `cause-calls: none (${r.reason ?? 'unknown'})${noteSuffix}`
  const head = `cause-calls: ${r.calls.length} spawn call(s) in ${r.windowFromIso}→${r.windowToIso} — ${composition(r.calls)}${noteSuffix}`
  const lines = r.calls.map(c => {
    const tag = [c.tool, c.subagentType && `subagent_type=${c.subagentType}`, c.model && `model=${c.model}`]
      .filter(Boolean).join(' ')
    return `  ${c.n}. ${c.iso}  ${tag}  ${c.input}`
  })
  return [head, ...lines].join('\n')
}

File: rust-core/crates/agentlens-core/src/burn/causing_tool_call.rs
//! Port of src/causingToolCall.ts — given a burn PEAK (a time window plus a session or
//! workspace), EVERY spawn tool-call that could have caused it, VERBATIM, numbered and
//! time-ordered. Not a single "nearest" call: a fork-storm is a SUSTAINED burst driven by
//! potentially many spawns, so reporting one is misleading.
//!
//! DELIBERATE ENGINE DIVERGENCE: the TS streams the NDJSON through DuckDB's read_json because
//! parsing ~2GB transcripts in JS is not viable. Rust parses NDJSON natively line-by-line
//! (for_each_json_line — the same streaming reader the log scanners use, one line buffer), so
//! the DuckDB dependency, its `maximum_object_size` tuning and the `duckdb-unavailable` reason
//! all disappear. The torn-line disclosure is kept and computed the same way: a line that fails
//! to parse is counted and reported, never silently dropped.

use serde_json::{Map, Value};
use std::path::{Path, PathBuf};

use crate::summarize::helpers::iso_from_ms;

/// Tools whose call fans out / spawns work — the FORK_STORM candidates.
pub const SPAWN_TOOLS: [&str; 4] = ["Task", "Agent", "Workflow", "SendMessage"];

const DEFAULT_WINDOW_MS: f64 = 15.0 * 60_000.0;
const DEFAULT_FORWARD_SLACK_MS: f64 = 90_000.0;
/// mtime is a coarse "this session was active near the peak" proxy — a session file spans time.
const MTIME_SLACK_MS: f64 = 3_600_000.0;
const MAX_FILES: usize = 8;
/// Claude Code truncates a project slug longer than this before appending its hash.
const SLUG_MAX_LEN: usize = 200;

/// projectSlugOf — every non-alphanumeric character becomes '-'; a bare slug (no separator)
/// passes through unchanged.
pub fn project_slug_of(path_or_slug: &str) -> String {
    let value = path_or_slug.trim();
    if value.is_empty() {
        return String::new();
    }
    if !value.contains('/') && !value.contains('\\') {
        return value.to_owned();
    }
    value.chars().map(|c| if c.is_ascii_alphanumeric() { c } else { '-' }).collect()
}

/// resolveProjectSlugs — the naive derivation is exact BELOW the truncation boundary; at or past
/// it the real directory name carries a hash this code deliberately does NOT reproduce (it was
/// measured to be none of md5/sha1/sha256/sha512 over the path or slug), so the name is read off
/// disk instead. Falls back to the naive slug when nothing matches.
pub fn resolve_project_slugs(path_or_slug: &str, roots: &[PathBuf]) -> Vec<String> {
    let naive = project_slug_of(path_or_slug);
    if naive.chars().count() < SLUG_MAX_LEN {
        return vec![naive];
    }
    let head: String = naive.chars().take(SLUG_MAX_LEN).collect();
    let mut found: Vec<String> = Vec::new();
    for root in roots {
        let Ok(rd) = std::fs::read_dir(root) else { continue };
        for name in rd.flatten().filter_map(|e| e.file_name().to_str().map(str::to_owned)) {
            let chars: Vec<char> = name.chars().collect();
            let hashed = chars.len() > SLUG_MAX_LEN && chars[SLUG_MAX_LEN] == '-' && name.starts_with(&head);
            if (name == naive || hashed) && !found.contains(&name) {
                found.push(name);
            }
        }
    }
    if found.is_empty() { vec![naive] } else { found }
}

pub struct CausingCallsOptions<'a> {
    pub at_ms: f64,
    pub session_id: Option<&'a str>,
    pub workspace: Option<&'a str>,
    pub jsonl_path: Option<&'a Path>,
    pub window_ms: Option<f64>,
    pub forward_slack_ms: Option<f64>,
    pub tools: Option<&'a [&'a str]>,
    pub projects_dirs: Vec<PathBuf>,
}

/// Stream a transcript line by line (ONE line buffer — transcripts reach GBs; the P5g lesson
/// that a whole-file read is the memory peak applies here too). `f` sees each raw line; the
/// caller decides what a parse failure means, which is what makes the torn-line count possible
/// (agentlens_logscan::for_each_json_line is crate-private AND hands over parsed maps, so a
/// line that failed to parse would be invisible to it).
fn for_each_line(path: &Path, mut f: impl FnMut(&str)) {
    use std::io::BufRead;
    let Ok(file) = std::fs::File::open(path) else { return };
    let mut reader = std::io::BufReader::new(file);
    let mut buf = Vec::new();
    loop {
        buf.clear();
        match reader.read_until(b'\n', &mut buf) {
            Ok(0) | Err(_) => return,
            Ok(_) => {}
        }
        while buf.last().is_some_and(|b| *b == b'\n' || *b == b'\r') {
            buf.pop();
        }
        if buf.is_empty() {
            continue;
        }
        f(&String::from_utf8_lossy(&buf));
    }
}

fn mtime_ms(p: &Path) -> Option<f64> {
    let md = std::fs::metadata(p).ok()?;
    md.modified()
        .ok()
        .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
        .map(|d| d.as_secs() as f64 * 1000.0 + d.subsec_nanos() as f64 / 1e6)
}

/// Candidate transcript file(s) for this peak, most-recently-modified first (capped).
fn resolve_transcripts(opts: &CausingCallsOptions) -> Vec<PathBuf> {
    if let Some(p) = opts.jsonl_path {
        return if std::fs::metadata(p).is_ok() { vec![p.to_path_buf()] } else { Vec::new() };
    }
    if let Some(sid) = opts.session_id.filter(|s| !s.is_empty()) {
        let mut out = Vec::new();
        for base in &opts.projects_dirs {
            let Ok(rd) = std::fs::read_dir(base) else { continue };
            for sub in rd.flatten().filter_map(|e| e.file_name().to_str().map(str::to_owned)) {
                let p = base.join(sub).join(format!("{sid}.jsonl"));
                if std::fs::metadata(&p).is_ok() {
                    out.push(p);
                }
            }
        }
        return out;
    }
    if let Some(ws) = opts.workspace.filter(|s| !s.is_empty()) {
        let slugs = resolve_project_slugs(ws, &opts.projects_dirs);
        let lo = opts.at_ms - opts.window_ms.unwrap_or(DEFAULT_WINDOW_MS) - MTIME_SLACK_MS;
        let hi = opts.at_ms + opts.forward_slack_ms.unwrap_or(DEFAULT_FORWARD_SLACK_MS) + MTIME_SLACK_MS;
        let mut cand: Vec<(PathBuf, f64)> = Vec::new();
        for base in &opts.projects_dirs {
            for slug in &slugs {
                let dir = base.join(slug);
                let Ok(rd) = std::fs::read_dir(&dir) else { continue };
                for name in rd.flatten().filter_map(|e| e.file_name().to_str().map(str::to_owned)) {
                    if !name.ends_with(".jsonl") {
                        continue;
                    }
                    let p = dir.join(&name);
                    let Some(mt) = mtime_ms(&p) else { continue }; // raced/unreadable — skip
                    if mt < lo || mt > hi {
                        continue;
                    }
                    cand.push((p, mt));
                }
            }
        }
        cand.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
        return cand.into_iter().take(MAX_FILES).map(|c| c.0).collect();
    }
    Vec::new()
}

/// causingToolCalls — every spawn tool-call in the peak window, numbered + time-ordered +
/// verbatim, or an empty list with an honest `reason` (never a fabricated call).
/// Wire shape: `{calls:[{n,iso,tool,subagentType,model,input,sessionId}], windowFromIso,
/// windowToIso, reason?, note?}`.
pub fn causing_tool_calls(opts: &CausingCallsOptions) -> Value {
    let window_ms = opts.window_ms.unwrap_or(DEFAULT_WINDOW_MS);
    let fwd = opts.forward_slack_ms.unwrap_or(DEFAULT_FORWARD_SLACK_MS);
    let window_from_iso = iso_from_ms(opts.at_ms - window_ms);
    let window_to_iso = iso_from_ms(opts.at_ms + fwd);
    let mut out = Map::new();

    let files = resolve_transcripts(opts);
    if files.is_empty() {
        let located = opts.jsonl_path.is_some() || opts.session_id.is_some_and(|s| !s.is_empty()) || opts.workspace.is_some_and(|s| !s.is_empty());
        out.insert("calls".into(), Value::Array(Vec::new()));
        out.insert("windowFromIso".into(), window_from_iso.into());
        out.insert("windowToIso".into(), window_to_iso.into());
        out.insert("reason".into(), if located { "no-transcript" } else { "no-locator" }.into());
        return Value::Object(out);
    }

    let tools: Vec<&str> = opts.tools.map(<[&str]>::to_vec).unwrap_or_else(|| SPAWN_TOOLS.to_vec());
    // (timestamp, tool, subagent_type, model, input, sessionId) per matching content block.
    let mut rows: Vec<(String, Value)> = Vec::new();
    let mut total_lines = 0u64;
    let mut with_type = 0u64;
    for file in &files {
        let session_from_file = file.file_stem().map(|s| s.to_string_lossy().into_owned()).unwrap_or_default();
        for_each_line(file, |line| {
            total_lines += 1;
            // The torn-line probe keys on `type`, NOT `timestamp`: measured over 482,993 real
            // records, `type` is missing from 0 while `timestamp` is missing from 16.9% (the
            // attachment / queue-operation / last-prompt records legitimately carry none), so a
            // timestamp-keyed probe would report a sixth of a healthy machine as unparseable.
            let Ok(v) = serde_json::from_str::<Value>(line) else { return };
            if v.get("type").is_some() {
                with_type += 1;
            }
            if v.get("type").and_then(Value::as_str) != Some("assistant") {
                return;
            }
            let Some(ts) = v.get("timestamp").and_then(Value::as_str) else { return };
            // String comparison, as the SQL does on ISO text.
            if ts < window_from_iso.as_str() || ts > window_to_iso.as_str() {
                return;
            }
            let Some(content) = v.get("message").and_then(|m| m.get("content")).and_then(Value::as_array) else { return };
            for block in content {
                if block.get("type").and_then(Value::as_str) != Some("tool_use") {
                    continue;
                }
                let Some(tool) = block.get("name").and_then(Value::as_str) else { continue };
                if !tools.contains(&tool) {
                    continue;
                }
                let input = block.get("input").cloned().unwrap_or(Value::Null);
                let mut r = Map::new();
                r.insert("iso".into(), ts.into());
                r.insert("tool".into(), tool.into());
                // json_extract_string returns NULL for a missing/non-string field.
                r.insert("subagentType".into(), input.get("subagent_type").and_then(Value::as_str).map_or(Value::Null, Value::from));
                r.insert("model".into(), input.get("model").and_then(Value::as_str).map_or(Value::Null, Value::from));
                // CAST(json_extract(...) AS VARCHAR) — the input serialized back, verbatim.
                r.insert("input".into(), input.to_string().into());
                r.insert("sessionId".into(), opts.session_id.map_or_else(|| session_from_file.clone(), str::to_owned).into());
                rows.push((ts.to_owned(), Value::Object(r)));
            }
        });
    }
    // ORDER BY timestamp ASC — stable, so equal timestamps keep file/line order.
    rows.sort_by(|a, b| a.0.cmp(&b.0));

    let torn = total_lines.saturating_sub(with_type);
    let note = (torn > 0).then(|| format!("{torn} line(s) unparseable and excluded."));
    out.insert(
        "calls".into(),
        Value::Array(
            rows.into_iter()
                .enumerate()
                .map(|(i, (_, mut r))| {
                    let obj = r.as_object_mut().expect("row object");
                    // `n` leads the wire object (the TS literal's key order).
                    let mut with_n = Map::new();
                    with_n.insert("n".into(), Value::from(i as i64 + 1));
                    for (k, v) in obj.iter() {
                        with_n.insert(k.clone(), v.clone());
                    }
                    Value::Object(with_n)
                })
                .collect(),
        ),
    );
    let empty = out.get("calls").and_then(Value::as_array).is_some_and(Vec::is_empty);
    out.insert("windowFromIso".into(), window_from_iso.into());
    out.insert("windowToIso".into(), window_to_iso.into());
    if empty {
        out.insert("reason".into(), "none-in-window".into());
    }
    if let Some(n) = note {
        out.insert("note".into(), n.into());
    }
    Value::Object(out)
}

/// composition — "Agent/general-purpose/opus×2, Workflow×2", most-frequent first.
pub fn composition(calls: &[Value]) -> String {
    let mut by: indexmap::IndexMap<String, f64> = indexmap::IndexMap::new();
    for c in calls {
        // `[tool, subagentType, model].filter(Boolean).join('/')` — null/'' both drop.
        let key = ["tool", "subagentType", "model"]
            .iter()
            .filter_map(|k| c.get(*k).and_then(Value::as_str).filter(|s| !s.is_empty()))
            .collect::<Vec<&str>>()
            .join("/");
        *by.entry(key).or_insert(0.0) += 1.0;
    }
    let mut ranked: Vec<(String, f64)> = by.into_iter().collect();
    ranked.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));
    ranked.into_iter().map(|(k, n)| format!("{k}×{}", crate::summarize::helpers::fmt_js_num(n))).collect::<Vec<String>>().join(", ")
}
